src/chain.ts

```diff
@@ -0,0 +1,169 @@
+import { Ucan, Capability, Fact } from "./types"
+import * as token from "./token"
+
+
+/**
+ * Represents a deeply verified chain of UCANs.
+ *
+ * These chains can actually be trees sometimes (if a ucan has >1 proofs).
+ */
+export class Chained {
+
+    // We need to keep the encoded version around to preserve the signature
+    private _encoded: string
+    private _decoded: Ucan<Chained>
+
+    constructor(encoded: string, decoded: Ucan<Chained>) {
+        this._encoded = encoded
+        this._decoded = decoded
+    }
+
+    /**
+     * Validate a UCAN chain from a given JWT-encoded UCAN.
+     * 
+     * This will validate
+     * - The encoding
+     * - The signatures (unless turned off in the `options`)
+     * - The UCAN time bounds (unless turned off in the `options`)
+     * - The audience from parent proof UCANs matching up with the issuer of child UCANs
+     * 
+     * @returns A promise of a deeply-validated, deeply-parsed UCAN.
+     * @throws If the UCAN chain can't be validated.
+     */
+    static async fromToken(encodedUcan: string, options?: token.ValidateOptions): Promise<Chained> {
+        const ucan = await token.validate(encodedUcan, options)
+
+        // parse proofs recursively
+        const proofs = await Promise.all(ucan.payload.prf.map(encodedPrf => Chained.fromToken(encodedPrf, options)))
+
+        // check sender/receiver matchups. A parent ucan's audience must match the child ucan's issuer
+        const incorrectProof = proofs.find(proof => proof.audience() !== ucan.payload.iss)
+        if (incorrectProof != null) {
+            throw new Error(`Invalid UCAN: Audience ${incorrectProof.audience()} doesn't match issuer ${ucan.payload.iss}`)
```

Comment:
assuming some way of doing attenuation checks are coming in the next one :eyes:

Reply:
Yep :100: